File: 351004/Kadushko/Lab1/restApi/src/main/java/com/example/restApi/services/CommentService.java
package com.example.restApi.services;

import com.example.restApi.dto.request.CommentRequestTo;
import com.example.restApi.dto.response.CommentResponseTo;
import com.example.restApi.exception.NotFoundException;
import com.example.restApi.mapper.CommentMapper;
import com.example.restApi.model.Comment;
import com.example.restApi.repository.CommentRepository;
import com.example.restApi.repository.IssueRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class CommentService {

    private final CommentRepository commentRepository;
    private final IssueRepository issueRepository;
    private final CommentMapper commentMapper;

    public CommentService(CommentRepository commentRepository,
                          IssueRepository issueRepository,
                          CommentMapper commentMapper) {
        this.commentRepository = commentRepository;
        this.issueRepository = issueRepository;
        this.commentMapper = commentMapper;
    }

    public List<CommentResponseTo> getAll() {
        return commentRepository.findAll().stream()
                .map(commentMapper::toDto)
                .collect(Collectors.toList());
    }

    public CommentResponseTo getById(Long id) {
        Comment comment = commentRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Comment not found with id: " + id));
        return commentMapper.toDto(comment);
    }

    public CommentResponseTo create(CommentRequestTo request) {
        if (!issueRepository.existsById(request.getIssueId())) {
            throw new NotFoundException("Issue not found with id: " + request.getIssueId());
        }
        Comment comment = commentMapper.toEntity(request);
        comment.setId(null);
        Comment saved = commentRepository.save(comment);
        return commentMapper.toDto(saved);
    }

    public CommentResponseTo update(CommentRequestTo request) {
        if (request.getId() == null) {
            throw new IllegalArgumentException("Comment ID is required for update");
        }
        Comment comment = commentRepository.findById(request.getId())
                .orElseThrow(() -> new NotFoundException("Comment not found with id: " + request.getId()));
        if (!issueRepository.existsById(request.getIssueId())) {
            throw new NotFoundException("Issue not found with id: " + request.getIssueId());
        }
        commentMapper.updateEntity(comment, request);
        Comment updated = commentRepository.save(comment);
        return commentMapper.toDto(updated);
    }

    public void delete(Long id) {
        if (!commentRepository.existsById(id)) {
            throw new NotFoundException("Comment not found with id: " + id);
        }
        commentRepository.deleteById(id);
    }
}

File: 351003/Soika/entities-app/src/main/kotlin/com/example/entitiesapp/service/CommentService.kt
package com.example.entitiesapp.service

import com.example.entitiesapp.dto.CommentRequestTo
import com.example.entitiesapp.dto.CommentResponseTo
import com.example.entitiesapp.exception.NotFoundException
import com.example.entitiesapp.exception.ValidationException
import com.example.entitiesapp.model.Comment
import com.example.entitiesapp.repository.CommentRepository
import com.example.entitiesapp.repository.StoryRepository
import org.springframework.stereotype.Service

@Service
class CommentService(
    private val commentRepository: CommentRepository,
    private val storyRepository: StoryRepository
) {

    private fun toEntity(dto: CommentRequestTo) =
        Comment(content = dto.content, storyId = dto.storyId)

    private fun toResponse(entity: Comment) =
        CommentResponseTo(id = entity.id!!, content = entity.content, storyId = entity.storyId)

    fun getAll(): List<CommentResponseTo> =
        commentRepository.findAll().map { toResponse(it) }

    fun getById(id: Long): CommentResponseTo =
        toResponse(
            commentRepository.findById(id)
                ?: throw NotFoundException("Comment with id=$id not found", 40404)
        )

    fun create(dto: CommentRequestTo): CommentResponseTo {
        if (storyRepository.findById(dto.storyId) == null)
            throw ValidationException("Story with id=${dto.storyId} does not exist", 40003)
        return toResponse(commentRepository.save(toEntity(dto)))
    }

    fun update(id: Long, dto: CommentRequestTo): CommentResponseTo {
        if (commentRepository.findById(id) == null)
            throw NotFoundException("Comment with id=$id not found", 40404)
        if (storyRepository.findById(dto.storyId) == null)
            throw ValidationException("Story with id=${dto.storyId} does not exist", 40003)
        return toResponse(commentRepository.update(id, toEntity(dto)))
    }

    fun delete(id: Long) {
        if (commentRepository.findById(id) == null)
            throw NotFoundException("Comment with id=$id not found", 40404)
        commentRepository.delete(id)
    }
}